File: src/main/java/me/zeroeightsix/kami/feature/module/render/Brightness.java
package me.zeroeightsix.kami.feature.module.render;

import io.github.fablabsmc.fablabs.api.fiber.v1.annotation.Setting;
import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.TickEvent;
import me.zeroeightsix.kami.feature.module.Module;
import me.zeroeightsix.kami.setting.SettingVisibility;

import java.util.Stack;
import java.util.function.Function;

/**
 * Created by 086 on 12/12/2017.
 * @see me.zeroeightsix.kami.mixin.client.MixinEntityRenderer
 */
@Module.Info(name = "Brightness", description = "Makes everything brighter!", category = Module.Category.RENDER)
public class Brightness extends Module {

    @Setting(name = "Transition")
    private boolean transition = true;
    @Setting(name = "Seconds")
    @SettingVisibility.Method("ifTransition")
    private @Setting.Constrain.Range(min = 0, max = 10) float seconds = 1;
    @Setting(name = "Mode")
    private Transition mode = Transition.SINE;
    
    public boolean ifTransition() {
        return transition;
    }

    private Stack<Float> transitionStack = new Stack<>();

    private static float currentBrightness = 0;
    private static boolean inTransition = false;

    private void addTransition(boolean isUpwards) {
        if (transition) {
            int length = (int) (seconds * 20);
            float[] values;
            switch (mode) {
                case LINEAR:
                    values = linear(length, isUpwards);
                    break;
                case SINE:
                    values = sine(length, isUpwards);
                    break;
                default:
                    values = new float[]{0};
                    break;
            }
            for (float v : values) {
                transitionStack.add(v);
            }

            inTransition = true;
        }
    }

    @Override
    public void onEnable() {
        super.onEnable();
        addTransition(true);
    }

    @Override
    public void onDisable() {
        setAlwaysListening(true);
        super.onDisable();
        addTransition(false);
    }

    @EventHandler
    private Listener<TickEvent.Client.InGame> updateListener = new Listener<>(event -> {
        if (inTransition) {
            if (transitionStack.isEmpty()) {
                inTransition = false;
                setAlwaysListening(false);
                currentBrightness = isEnabled() ? 1 : 0;
            } else {
                currentBrightness = transitionStack.pop();
            }
        }
    });

    private float[] createTransition(int length, boolean upwards, Function<Float, Float> function) {
        float[] transition = new float[length];
        for (int i = 0; i < length; i++) {
            float v = function.apply(((float) i / (float) length));
            if (upwards) v = 1 - v;
            transition[i] = v;
        }
        return transition;
    }

    private float[] linear(int length, boolean polarity) { // length of 20 = 1 second
        return createTransition(length, polarity, d -> d);
    }

    private float sine(float x) { // (sin(pi*x-(pi/2)) + 1) / 2
        return ((float) Math.sin(Math.PI * x - Math.PI / 2) + 1) / 2;
    }

    private float[] sine(int length, boolean polarity) {
        return createTransition(length, polarity, this::sine);
    }

    public static float getCurrentBrightness() {
        return currentBrightness;
    }

    public static boolean isInTransition() {
        return inTransition;
    }

    public static boolean shouldBeActive() {
        return isInTransition() || currentBrightness == 1; // if in transition or enabled
    }

    public enum Transition {
        LINEAR, SINE

    }
}

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/Brightness.kt
package me.zeroeightsix.kami.feature.module

import io.github.fablabsmc.fablabs.api.fiber.v1.annotation.Setting
import io.github.fablabsmc.fablabs.api.fiber.v1.annotation.Setting.Constrain
import me.zero.alpine.listener.EventHandler
import me.zero.alpine.listener.Listener
import me.zeroeightsix.kami.event.TickEvent.InGame
import kotlin.math.sin

@Module.Info(name = "Brightness", description = "Makes everything brighter!", category = Module.Category.RENDER)
object Brightness : Module() {
    @Setting(name = "Transition length")
    private var seconds: @Constrain.Range(min = 0.0, max = 10.0, step = 0.1) Float = 1f

    @Setting(comment = "The factor by which to multiply the final brightness. A factor of 1 indicates the normal night vision brightness.")
    private var factor: @Constrain.Range(min = 0.1, max = 2.0, step = 0.02) Float = 1f

    private var currentBrightness = 0f

    private var transitionStack: ArrayDeque<Float>? = null
    
    override fun onEnable() {
        val (min, max) = if (transitionStack != null) Pair(currentBrightness, 1f) else Pair(0f, 1f)
        transitionStack = createTransition((seconds * 20).toInt(), true, min, max)
        super.onEnable()
    }

    override fun onDisable() {
        val (min, max) = if (transitionStack != null) Pair(0f, currentBrightness) else Pair(0f, 1f)
        transitionStack = createTransition((seconds * 20).toInt(), false, min, max)
        alwaysListening = true
        super.onDisable()
    }

    @EventHandler
    private val updateListener = Listener<InGame>(
        {
            transitionStack?.let {
                val b = it.removeLastOrNull()
                // If there are no more entries, the transition has ended.
                // Stop listening, and set the final brightness.
                if (b == null) {
                    alwaysListening = false
                    currentBrightness = if (enabled) 1f else 0f
                    return@Listener
                }
                currentBrightness = b
            }
        })

    private fun createTransition(length: Int, upwards: Boolean, min: Float = 0f, max: Float = 1f): ArrayDeque<Float> {
        val queue = ArrayDeque<Float>(length)
        val fl = length.toFloat()
        val height = max - min
        (0 until length).forEach {
            var v = min + sine(it.toFloat() / fl) * height
            if (upwards) v = max - v
            queue.add(v)
        }
        return queue
    }

    private fun sine(x: Float): Float { // (sin(pi*x-(pi/2)) + 1) / 2
        return (sin(Math.PI * x - Math.PI / 2).toFloat() + 1) / 2
    }

    @JvmStatic
    fun getCurrentBrightness() = currentBrightness * factor

    @JvmStatic
    fun shouldBeActive(): Boolean {
        // != null if in transition
        return enabled || transitionStack != null
    }
}